libraries/AP_InertialSensor/AP_InertialSensor_MPU6000.cpp

```diff
@@ -313,27 +316,33 @@ bool AP_InertialSensor_MPU6000::_init()
 
     bool success = _hardware_init();
 
-#if MPU6000_DEBUG
-    _dump_registers();
-#endif
-
     return success;
 }
 
 void AP_InertialSensor_MPU6000::_fifo_reset()
 {
     uint8_t user_ctrl = _master_i2c_enable?BIT_USER_CTRL_I2C_MST_EN:0;
+    _dev->set_speed(AP_HAL::Device::SPEED_LOW);
+    _register_write(MPUREG_FIFO_EN, 0);
     _register_write(MPUREG_USER_CTRL, user_ctrl);
     _register_write(MPUREG_USER_CTRL, user_ctrl | BIT_USER_CTRL_FIFO_RESET);
     _register_write(MPUREG_USER_CTRL, user_ctrl | BIT_USER_CTRL_FIFO_EN);
+    _register_write(MPUREG_FIFO_EN, BIT_XG_FIFO_EN | BIT_YG_FIFO_EN |
+                    BIT_ZG_FIFO_EN | BIT_ACCEL_FIFO_EN | BIT_TEMP_FIFO_EN);
+    hal.scheduler->delay_microseconds(1);
+    _dev->set_speed(AP_HAL::Device::SPEED_HIGH);
 }
 
 void AP_InertialSensor_MPU6000::_fifo_enable()
 {
+    uint8_t user_ctrl = _master_i2c_enable?BIT_USER_CTRL_I2C_MST_EN:0;
```

Comment:
It's still crashing the FIFO using i2c master mode. We need to store entirely the control user regiter or as default adding a regiter read again. Adding this maybe can help work with high speed MPU samples without adding any other hardcoded strategy. Using FIFO the copter fly smooth directly.